Network/HTTP/Req.hs

```diff
@@ -1297,13 +1301,24 @@ instance HttpBody ReqBodyUrlEnc where
 
 -- | An opaque monoidal value that allows to collect URL-encoded parameters
 -- to be wrapped in 'ReqBodyUrlEnc'.
+--
+-- You can inspect its content with 'toList'.
```

Comment:
`IsList` removed the guarantees that the smart constructor provides. It can be used not only for inspection but also for constructing invalid values.

Reply:
OK. `IsList` instance was an overkill.

What do you think of adding `formToList :: FormUrlEncodedParam -> [(Text, Maybe Text)]` instead?

Reply:
It looks like the most logical place for this functionality is in the `QueryParam` type class. Along with `queryParam` there could be `paramToList`.

Reply:
That would be great! I'll try that.

Reply:
Actually, what is the motivation for having this kind of function?

Reply:
Actually, what is the motivation for having this kind of function?

Reply:
It's mainly for tests.

I wrote a client SDK for a Web API with req as the backend HTTP client. Some endpoints of the Web API required `application/x-www-form-urlencoded` body. So, my client SDK needed to translate a Haskell object (supplied by the user of the SDK) into the HTTP body. That means I wrote a function like the following.

```haskell
buildBody :: InputByUser -> FormUrlEncodedParam
```

Because the algorithm of `buildBody` was a little complicated, I wanted to write unit tests on it. However, I couldn't test `buildBody` directly, because `FormUrlEncodedParam` didn't have any inspection interface.

I also considered building `Request` by `req'` function and testing it. However, the [`RequestBody`](https://hackage.haskell.org/package/http-client-0.7.11/docs/Network-HTTP-Client.html#t:RequestBody) type was a sum type, and I had no idea which variant was the right one (I also think that should be a hidden implementation detail). In addition, because `RequestBody` can be a streaming content, it would need some work to get the form body.

If `FormUrlEncodedParam` had some inspection interface in the first place, it would be so easy to write the test on `buildBody`.